src/autoapp/Projection/GSTVideoOutput.cpp

```diff
@@ -0,0 +1,260 @@
+    gst_object_unref(vid_pipeline);
+    gst_object_unref(vid_src);
+}
+
+gboolean GSTVideoOutput::bus_callback(GstBus */* unused*/, GstMessage *message, gpointer *ptr) {
+    gchar *debug;
+    GError *err;
+    gchar *name;
+
+    switch (GST_MESSAGE_TYPE(message)) {
+
+    case GST_MESSAGE_ERROR:
+        gst_message_parse_error(message, &err, &debug);
+        OPENAUTO_LOG(info) << "[GSTVideoOutput] Error "<< err->message;
+        g_error_free(err);
+        g_free(debug);
+        break;
+
+    case GST_MESSAGE_WARNING:
+        gst_message_parse_warning(message, &err, &debug);
+        OPENAUTO_LOG(info) << "[GSTVideoOutput] Warning "<<err->message<<" | Debug "<< debug;
+
+        name = (gchar *) GST_MESSAGE_SRC_NAME(message);
+
+        OPENAUTO_LOG(info) << "[GSTVideoOutput] Name of src "<< name ? name : "nil";
+        g_error_free(err);
+        g_free(debug);
+
+        break;
+
+    case GST_MESSAGE_EOS:
+        OPENAUTO_LOG(info) << "[GSTVideoOutput] End of stream";
+        break;
+
+    case GST_MESSAGE_STATE_CHANGED:
+        break;
+    default:
+        break;
+    }
+
+    return TRUE;
+}
+bool GSTVideoOutput::open()
+{
+     GstElement *capsfilter = gst_bin_get_by_name(GST_BIN(vid_pipeline), "mycapsfilter");
+    GstPad *convert_pad = gst_element_get_static_pad(capsfilter, "sink");
+    gst_pad_add_probe (convert_pad,GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM,convert_probe, this, NULL);
+    gst_element_set_state(vid_pipeline, GST_STATE_PLAYING);
+    return true;
+}
+
+GstPadProbeReturn GSTVideoOutput::convert_probe(GstPad *pad, GstPadProbeInfo *info, void *user_data){
+    GstEvent *event = GST_PAD_PROBE_INFO_EVENT(info);
+    if (GST_PAD_PROBE_INFO_TYPE(info) & GST_PAD_PROBE_TYPE_EVENT_DOWNSTREAM) {
+        if (GST_EVENT_TYPE (event) == GST_EVENT_SEGMENT) {
+            GstCaps *caps  = gst_pad_get_current_caps(pad);
+            if(caps != NULL){
+                GstVideoInfo *vinfo = gst_video_info_new ();
+                gst_video_info_from_caps (vinfo, caps);
+                OPENAUTO_LOG(info) << "[GSTVideoOutput] Video Width: "<<vinfo->width;
+                OPENAUTO_LOG(info) << "[GSTVideoOutput] Video Height: "<<vinfo->height;
+
+
+            }
+            return GST_PAD_PROBE_REMOVE;
+        }
+    }
+    return GST_PAD_PROBE_OK;
+}
+bool GSTVideoOutput::init()
+{
+
+    OPENAUTO_LOG(info) << "[GSTVideoOutput] init";
+    emit startPlayback();
+    return true;
+}
+void GSTVideoOutput::write(uint64_t timestamp, const aasdk::common::DataConstBuffer& buffer)
+{
+    GstBuffer * buffer_ = gst_buffer_new_and_alloc(buffer.size);
+    gst_buffer_fill(buffer_, 0, buffer.cdata, buffer.size);
+    int ret = gst_app_src_push_buffer((GstAppSrc *) vid_src, buffer_);
+    if (ret != GST_FLOW_OK) {
+        OPENAUTO_LOG(info)<<"[GSTVideoOutput] push buffer returned "<< ret <<" for "<< buffer.size <<"bytes";
+    }
+
+}
+
+void GSTVideoOutput::onStartPlayback()
+{
+    if(activeCallback_ != nullptr)
```

Comment:
since this has support for signal/slots, would we just want to connect  to the slot instead of using the callback (kinda like how its done for qt video in the service factory)

Reply:
Tried it - might have tried it badly, but resulted in a weird QML glcontext crash. Implementing more in the style of OMXProjection worked fine.